cmake: Add `CheckLinkerSupportsPIE` module

hebasto opened the pull request:
This new module is a wrapper around CMake's `CheckPIESupported` module that fixes an upstream bug.

See: https://gitlab.kitware.com/cmake/cmake/-/issues/26463.

Fixes https://github.com/bitcoin/bitcoin/issues/30771.

hebasto commented:
cc @vasild

hebasto commented:
The upstream [fix](https://gitlab.kitware.com/cmake/cmake/-/merge_requests/10034) has been mentioned in the comment.

fanquake commented:
> Fixes an upstream bug
> Enhances robustness

Wondering if in this case given CMakes tooling has not only been buggy, but then is also apparently not robust enough for production use without wrapping it in more layers of abstraction, maybe we should just revert to (as we did in Autotools) something like using the flags we want directly.

vasild reviewed (approved):
ACK d3f42fa08fc385752881afa5740f40287cfb4d8b

Without this patch I get:

```
CMake Warning at CMakeLists.txt:195 (message):
  PIE is not supported at link time: PIE (CXX): Change Dir:
  '/tmp/build/clang20-fuzz0/CMakeFiles/CMakeScratch/TryCompile-8aYRFx'
...
  ld: error: relocation R_X86_64_32S cannot be used against local symbol;
  recompile with -fPIC
```

with this patch - no errors.

hebasto commented:
> > Enhances robustness
> 
> Can you explain this further? Why is CMakes own check not "robust" enough to use directly? Is this another bug/something that should be improved upstream?

Sure thing!

(Almost?) every CMake check internally uses the [`try_compile()`](https://cmake.org/cmake/help/latest/command/try_compile.html) command, whose behaviour, in turn, depends on the [`CMAKE_TRY_COMPILE_TARGET_TYPE`](https://cmake.org/cmake/help/latest/variable/CMAKE_TRY_COMPILE_TARGET_TYPE.html) variable:  
1. The default value, `EXECUTABLE`, enables both compiler and linker checks.  
2. The `STATIC_LIBRARY` value enables only compiler checks.  

To mimic Autotools' behaviour, we [disabled](https://github.com/bitcoin/bitcoin/blob/d3f42fa08fc385752881afa5740f40287cfb4d8b/cmake/module/CheckSourceCompilesAndLinks.cmake#L9-L10) linker checks by setting `CMAKE_TRY_COMPILE_TARGET_TYPE` to `STATIC_LIBRARY` globally (perhaps not the best design). This effectively separates the entire CMake script into regions where `CMAKE_TRY_COMPILE_TARGET_TYPE` is:  
- unset  
- set to `STATIC_LIBRARY`  
- set to `EXECUTABLE`  

The following line:https://github.com/bitcoin/bitcoin/blob/d3f42fa08fc385752881afa5740f40287cfb4d8b/cmake/module/CheckLinkerSupportsPIE.cmake#L8-L9
does not change the current behaviour because, at this point, `CMAKE_TRY_COMPILE_TARGET_TYPE` is unset, which is equivalent to the `EXECUTABLE` value. However, with this line, it is safe to move the `check_linker_supports_pie()` command around.

fanquake commented:
This seems very fragile and unintuitive, and the fact that this could silently break at any point is not documented in any way. I don't think other bad design decisions should lead to us having to write even more boilerplate code to fix things that should "just work" (minus the upstream bugs).

theuni commented:
> This seems very fragile and unintuitive, and the fact that this could silently break at any point is not documented in any way. I don't think other bad design decisions should lead to us having to write even more boilerplate code to fix things that should "just work" (minus the upstream bugs).

Agreed. I forgot that we set `CMAKE_TRY_COMPILE_TARGET_TYPE` globally. And even worse, it's buried in a module. If that upsets CMake internal tests, I think we should undo that.

Could we invert this logic? ie turn `check_cxx_source_*` and friends to functions and set `CMAKE_TRY_COMPILE_TARGET_TYPE` to static in local scopes?

Concept ACK to wrapping for the upstream bug though. That's unfortunate :(

fanquake commented:
What is the status of this

hebasto committed:
cmake: Add `CheckLinkerSupportsPIE` module

This new module serves as a wrapper around CMake's `CheckPIESupported`
module and addresses an upstream bug:
- https://gitlab.kitware.com/cmake/cmake/-/issues/26463
- https://gitlab.kitware.com/cmake/cmake/-/merge_requests/10034

hebasto commented:
> I forgot that we set `CMAKE_TRY_COMPILE_TARGET_TYPE` globally. And even worse, it's buried in a module. If that upsets CMake internal tests, I think we should undo that.

Undone in https://github.com/bitcoin/bitcoin/pull/31662.

> What is the status of this?

Updated and ready for review.

hebasto committed:
cmake: Refer to the configure log instead of printing PIE test error

This change improves the user experience on systems where the toolchain
does not support PIE.

theuni commented:
The real fix here is in #31662, no?

And do we want to be just silently building without pie? Should we error unless hardening is disabled?

hebasto commented:
@theuni 

> The real fix here is in #31662, no?

The bug described in https://github.com/bitcoin/bitcoin/issues/30771#issue-2497352810 / https://gitlab.kitware.com/cmake/cmake/-/issues/26463 is fixed in the first commit:https://github.com/bitcoin/bitcoin/blob/65a0920ca6b11b4250b5cdcf174eaa9ea55fa6ed/cmake/module/CheckLinkerSupportsPIE.cmake#L8-L16

Changes from https://github.com/bitcoin/bitcoin/pull/31662 do not affect the `CheckLinkerSupportsPIE` module because the `CMAKE_TRY_COMPILE_TARGET_TYPE` variable was not modified prior to the module's usage.

> And do we want to be just silently building without pie? Should we error unless hardening is disabled?

IIRC, during the development of the staging branch, we had a discussion on this topic and decided to warn the user when PIE is not available. Since this PR does not alter that behaviour, could we continue this discussion in a dedicated issue or PR?

Besides, it does not affect our Tier 1 (release) builds or Tier 2 builds (documented or tested in the CI).

theuni reviewed (approved):
utACK 81c174e3186efae084829dcde314b081cad3d3cb.

Nice work helping to get this fixed upstream. Makes sense for us to carry the workaround for a few years. No opinion on the second commit.

vasild reviewed (approved):
ACK 81c174e3186efae084829dcde314b081cad3d3cb

I re-checked this. Without this PR I get the warning as mentioned in https://github.com/bitcoin/bitcoin/pull/31359#pullrequestreview-2465699855 and with this PR there is no warning. Further, without this PR, there is no `-fPIC` or `-fPIE` when compiling. With this PR `-fPIC` is added (not `-fPIE`).

hebasto commented on cmake/module/CheckLinkerSupportsPIE.cmake:
```diff
@@ -0,0 +1,27 @@
+    set(CMAKE_POSITION_INDEPENDENT_CODE ON PARENT_SCOPE)
+  elseif(NOT WIN32)
+    # The warning is superfluous for Windows.
+    message(WARNING "PIE is not supported at link time. See the configure log for details.")
```
It might depend on the used CMake version: https://github.com/bitcoin/bitcoin/blob/048ef98626b9ed62355923f089ad5cc5b6a898a3/ci/test/GetCMakeLogFiles.cmake#L5-L9

vasild replied:
Ok, not so important. If this recurs then maybe move that snippet to a location where it can be reused by `ci/test/GetCMakeLogFiles.cmake` and by `cmake/module/CheckLinkerSupportsPIE.cmake`, so they can use `${log_files}`.
